zmq4.go

```diff
@@ -32,6 +35,9 @@ type Socket interface {
 	// Dial connects a remote endpoint to the Socket.
 	Dial(ep string) error
 
+	// Dial connects a remote endpoint to the Socket.
+	DialContext(ctx context.Context, ep string) error
```

Comment:
the doc string of the `Dial` method should be updated to reflect the implied semantics.

the `csocket` implementation (for integration w/ cgo-zmq4) should be updated as well.

alternatively, we could just merge the 2 methods together, documenting that the `context.Context` would be wrapped with the socket's configured timeout.

I think I'd err on the latter.
the `csocket` implementation would still need to be updated, though :)